# HTTP Request/Response Cycle - Lab

## Introduction 

In this lab, we'll make use of the `requests` module commands and properties seen in the previous lesson, to extract information for a web service called **"Open Notify"** to access NASA's space data. 

## Objectives

You will be able to:

* Explain the HTTP request/response cycle
* List the status codes of responses and their meanings
* Obtain and interpret status codes from responses
* Make HTTP GET and POST requests in python using the `requests` library

## Open Notify 

[Open Notify](http://open-notify.org/)  is an an open source project to provide a simple programming interface for some of NASA’s awesome data. This takes live raw data from NASA's systems and turn them into APIs related to space and spacecraft. We can access the following information from open notify. 

* Current Location of the International Space Station

* Overhead Pass Predictions for the International Space Station

* Number of People in Space
    
### API endpoints

Open Notify has several API endpoints. 
>An endpoint is a server route that is used to retrieve different data from the API. 

For example, the `/comments` endpoint on the Reddit API might retrieve information about comments, whereas the `/users` endpoint might retrieve data about users. To access them, you would add the endpoint to the base url of the API.

For the OpenNotify API, we have following end points 

1. Current Location of the International Space Station `/iss-now.json`
2. Overhead Pass Predictions for the International Space Station `/iss-pass.json`    
3. Number of People in Space `/astros.json`

The `.json` extension simply tells us that the data is being returned in a JSON format.

In this lab, we'll be querying this API to retrieve live data about the International Space Station (ISS). Details on OpenNofity, endpoints, syntax, and the services it offers can be viewed [Here](http://open-notify.org/Open-Notify-API/)

![](images/iss.jpg)

### Current location of International Space Station

The first endpoint we'll look at on Open Notify is the ` iss-now.json` endpoint (current location of international space station). This endpoint gets the current latitude and longitude of the International Space Station.  Perform the following tasks 
* Make a get request to get the latest position of the international space station from the opennotify api's `iss-now` endpoint at http://api.open-notify.org/iss-now.json
* Check the status code of the response
* Interpret the returned code

In [1]:
# Your Code Here
import requests

In [3]:
# Your comments 
resp = requests.get('http://api.open-notify.org/iss-now.json')

resp.status_code == requests.codes.ok

True

* Print the contents of the response and identify its current location

In [4]:
# Your Code Here
print (resp.text)

{"message": "success", "iss_position": {"longitude": "68.3248", "latitude": "-39.6697"}, "timestamp": 1630935613}


# Interpret your results using the API
The iss is at latitude -39.6697 and  longitude 68.3248
https://www.google.co.uk/maps/place/39%C2%B040'10.9%22S+68%C2%B019'29.3%22E/@-39.6697,32.4654265,3z/data=!4m5!3m4!1s0x0:0x0!8m2!3d-39.6697!4d68.3248

### Check the next pass of International space station for a given location

Let's repeat the above for the second endpoint `iss-pass.json`. This end point is used to query the next pass of the space station on a given location. Let's just run as above and record your observations.

In [5]:
# Your Code Here
response = requests.get("http://api.open-notify.org/iss-pass.json")
print(response.status_code)

400



# Your comments 

400 is a bad request response due to a client error

So clearly there is something wrong as we had a 400 response. This is how you should always test your responses for validity. 

If we look at the documentation for the Open Notify API, we see that the ISS Pass endpoint requires two parameters.

> The ISS Pass endpoint returns when the ISS will next pass over a given location on earth. In order to compute this, we need to pass the coordinates of the location to the API. We do this by passing two parameters -- latitude and longitude.

We can do this by adding an optional keyword argument, `params`, to our request. In this case, there are two parameters we need to pass:

* `lat` -- The latitude of the location we want.
* `lon` -- The longitude of the location we want.

Perform the following tasks :
* Set parameters to reflect the lat and long of New York  (40.71, -74)
* Send a get request to Open Notify passing in the lat long parameters as k:v pairs in a dictionary
* Check the status code and interpret
* Print the header information and the returned content

In [7]:
# Your Code Here
latLong = {'lat': 51.5, 'lon': 0.1} 
response = requests.get("http://api.open-notify.org/iss-pass.json", params = latLong)
print(response.status_code)

200


In [8]:
# Check the API and interpret your results - when will ISS pass over London next ?

print(dict(response.headers))
print(response.text)

{'Server': 'nginx/1.10.3', 'Date': 'Mon, 06 Sep 2021 13:48:49 GMT', 'Content-Type': 'application/json', 'Content-Length': '516', 'Connection': 'keep-alive', 'Via': '1.1 vegur'}
{
  "message": "success", 
  "request": {
    "altitude": 100, 
    "datetime": 1630936129, 
    "latitude": 51.5, 
    "longitude": 0.1, 
    "passes": 5
  }, 
  "response": [
    {
      "duration": 573, 
      "risetime": 1630970572
    }, 
    {
      "duration": 651, 
      "risetime": 1630976312
    }, 
    {
      "duration": 658, 
      "risetime": 1630982116
    }, 
    {
      "duration": 654, 
      "risetime": 1630987926
    }, 
    {
      "duration": 587, 
      "risetime": 1630993742
    }
  ]
}



In [9]:

import time

In [10]:

over = response.json()['response'][0]['risetime']
print(f"ISS will next pass over London at {time.ctime(over)}")

ISS will next pass over London at Tue Sep  7 00:22:52 2021


### Finding the number of people in space

Open Notify has one more API endpoint, `/astros.json`. It tells you how many people are currently in space. The format of the responses can be studied [HERE](http://open-notify.org/Open-Notify-API/People-In-Space/).

Read the above documentation and perform the following tasks:

* Get the response from astros.json endpoint
* Count how many people are currently in space
* List the names of people currently in space.

In [11]:
# Your Code Here
astroReq = requests.get("http://api.open-notify.org/astros.json")
astros = astroReq.json()
astros

{'people': [{'name': 'Mark Vande Hei', 'craft': 'ISS'},
  {'name': 'Oleg Novitskiy', 'craft': 'ISS'},
  {'name': 'Pyotr Dubrov', 'craft': 'ISS'},
  {'name': 'Thomas Pesquet', 'craft': 'ISS'},
  {'name': 'Megan McArthur', 'craft': 'ISS'},
  {'name': 'Shane Kimbrough', 'craft': 'ISS'},
  {'name': 'Akihiko Hoshide', 'craft': 'ISS'},
  {'name': 'Nie Haisheng', 'craft': 'Tiangong'},
  {'name': 'Liu Boming', 'craft': 'Tiangong'},
  {'name': 'Tang Hongbo', 'craft': 'Tiangong'}],
 'number': 10,
 'message': 'success'}

In [17]:
# Interpret the Results - How many people are in space and what are their names 

names = [n['name'] for n in astros['people']]
print(f"There are {astros['number']} astronauts currently in space")
         
print(f"Their names are {', '.join(names)}")


There are 10 astronauts currently in space
Their names are Mark Vande Hei, Oleg Novitskiy, Pyotr Dubrov, Thomas Pesquet, Megan McArthur, Shane Kimbrough, Akihiko Hoshide, Nie Haisheng, Liu Boming, Tang Hongbo


## Summary 

In this lesson, we saw how we can use request and response methods to query an Open API. We also saw how to look at the contents returned with the API calls and how to parse them. Next, we'll look at connecting to APIs which are not OPEN, i.e. we would need to pass in some authentication information and filter the results. 